### ISSUE TITLE
[ERROR] occlum-pal: The init process exit with code: 139

### ISSUE THREAD

**Issue by StanPlatinum** [occlum/occlum#364]
# Describe the bug

[ERROR] occlum-pal: The init process exit with code: 139 (line 58, file src/pal_api.c)
[ERROR] occlum-pal: Failed to run the init process: EINVAL (line 125, file src/pal_api.c)
thread '[ERROR] occlum-pal: Failed to do ECall: occlum_ecall_broadcast_interrupts with error code 0x2002: Invalid enclave identification. (line 27, file src/pal_interrupt_thread.c)
main' panicked at 'Occlum image initialization failed: -1', src/bin/occlum_exec_server.rs:95:32
stack backtrace:
start_server failed Failed to launch server
Error: -1
Makefile:75: recipe for target 'pretest' failed
make[1]: *** [pretest] Error 1
Makefile:56: recipe for target 'test' failed
make: *** [test] Error 2

# To reproduce

Build the container,
1. docker run -it --device /dev/isgx occlum/occlum:0.21.0-ubuntu18.04
2. cd /opt/intel/sgxsdk/SampleCode/SampleEnclave && make && ./app

The sample enclave works fine.

Inside the container,
1. git clone https://github.com/occlum/occlum .
2. make submodule
3. make
4. make test

# Expected behavior

Error at `make test` step.

# Logs

root@2e2abb769a53:~/occlum# make test
/root/occlum/build/test initialized as an Occlum instance
BUILD TEST => client
CC <= /root/occlum/build/test/obj/client/main.o
LINK => /root/occlum/build/test/image/bin/client
DONE

...

BUILD TEST => unix_socket_throughput
CC <= /root/occlum/build/test/obj/unix_socket_throughput/main.o
LINK => /root/occlum/build/test/image/bin/unix_socket_throughput
DONE
Enclave sign-tool: /opt/occlum/sgxsdk-tools/bin/x64/sgx_sign
Enclave sign-key: /root/occlum/etc/template/Enclave.pem
SGX mode: HW
Building new image...
[+] Home dir is /root
[-] Open token file /root/enclave.token error! Will create one.
[+] Saved updated launch token!
[+] Init Enclave Successful 79658758438914!
Generate the SEFS image successfully
EXPORT => OCCLUM_CONF_FILE_MAC = 98-81-51-b1-3c-cb-eb-2f-bc-a4-89-ab-a8-9b-33-34
Building the initfs...
[+] Home dir is /root
[+] Open token file success! 
[+] Token file valid!
[+] Init Enclave Successful 79727477915650!
Generate the SEFS image successfully
Building libOS...
Signing the enclave...

...

tcs_num 32, tcs_max_num 32, tcs_min_pool 1
The required memory is 523452416B.
The required memory is 0x1f334000, 511184 KB.
Succeed.
Built the Occlum image and enclave successfully
[ERROR] occlum-pal: The init process exit with code: 139 (line 58, file src/pal_api.c)
[ERROR] occlum-pal: Failed to run the init process: EINVAL (line 125, file src/pal_api.c)
thread '[ERROR] occlum-pal: Failed to do ECall: occlum_ecall_broadcast_interrupts with error code 0x2002: Invalid enclave identification. (line 27, file src/pal_interrupt_thread.c)
main' panicked at 'Occlum image initialization failed: -1', src/bin/occlum_exec_server.rs:95:32
stack backtrace:
start_server failed Failed to launch server
Error: -1
Makefile:75: recipe for target 'pretest' failed
make[1]: *** [pretest] Error 1
Makefile:56: recipe for target 'test' failed
make: *** [test] Error 2

# Environment

- HW: [SGX2 - NUC7PJYH box] 
- OS: [Ubuntu18.04]
- Occlum version: [0.21.0]

**Comment by guzongmin** [comment]
@StanPlatinum I guess the fsgs driver does not work for you. If you do not install the fsgsbase driver, please follow the README to install it, If you are using a new Ubuntu kernel, please downgrade it to the orignal one (or build your own 5.11 kernel), because the latest Ubuntu driver forbidden the driver. 

 "Install enable_rdfsbase kernel module, which enables Occlum to use rdfsbase-family instructions in enclaves."

**Comment by StanPlatinum** [comment]
@guzongmin Good to see you, Zongmin!

Thanks for the reply.
I'm using Linux 5.9 where the fsgsbase feature is enabled.

```
LD_SHOW_AUXV=1 /bin/true | grep AT_HWCAP2                    ✔ 
AT_HWCAP2:       0x2
```
Do I still need to install the fsgsbase driver? Anyway I'm going to give it a try, though. Will let you know the results.

BTW, is the driver the following link?
https://github.com/occlum/enable_rdfsbase

Thanks!

**Comment by StanPlatinum** [comment]
After I installed the fsgs driver at https://github.com/occlum/enable_rdfsbase, the same error still exists...

**Comment by guzongmin** [comment]
@StanPlatinum I am using Ubuntu 5.4.0-42-generic now. Can you please try: one of the hello world demo with occlum gdb to help me get more information?

**Comment by guzongmin** [comment]
@StanPlatinum Could you please enable trace by setting environment variable "OCCLUM_LOG_LEVEL=trace"? It would print more information out.

**Comment by StanPlatinum** [comment]
@guzongmin Sorry for the late reply. Thanks! I changed to another machine then the error gone. It seems the SGX2 box machine has weird behavior when running Occlum... 

Anyway, I guess I can still do the hello demo and the tracing and let you know more details. But close the issue for now. Thanks again! Will catch you later if I have something new...

**Comment by guzongmin** [comment]
> @guzongmin Sorry for the late reply. Thanks! I changed to another machine then the error gone. It seems the SGX2 box machine has weird behavior when running Occlum...
> 
> Anyway, I guess I can still do the hello demo and the tracing and let you know more details. But close the issue for now. Thanks again! Will catch you later if I have something new...

If you are using SGX 2 with Intel in-tree driver, then the SGX device node name changed. You need to change the docker image launch commnd.

**Comment by StanPlatinum** [comment]
> > @guzongmin Sorry for the late reply. Thanks! I changed to another machine then the error gone. It seems the SGX2 box machine has weird behavior when running Occlum...
> > Anyway, I guess I can still do the hello demo and the tracing and let you know more details. But close the issue for now. Thanks again! Will catch you later if I have something new...
> 
> If you are using SGX 2 with Intel in-tree driver, then the SGX device node name changed. You need to change the docker image launch command.

I am using the out-of-tree driver. And I have installed the EPID-based attestation library/driver, not the DCAP version. The driver name is `isgx`, which is specified in the docker image launch command. 

Here I mention @ya0guang who also get interested in the issue...

@guzongmin I will post the output of hello demo later. Thanks!

**Comment by StanPlatinum** [comment]
logs,

```
root@24da375af3dd:~/demos/hello_c/occlum_instance# OCCLUM_LOG_LEVEL=trace occlum run /bin/hello_world
[2021-03-20T01:01:23.731Z][TRACE][T0][#0] env_checked from env untrusted: []
[2021-03-20T01:01:23.732Z][TRACE][T0][#0] env_merged = ["OCCLUM=yes"]  (default env and untrusted env)
[2021-03-20T01:01:23.732Z][DEBUG][T0][#0] lookup_inode: cwd: "/", path: "/bin/init"
[2021-03-20T01:01:23.769Z][TRACE][T0][#0] PT_INTERP offset = 568, count = 25
[2021-03-20T01:01:23.769Z][TRACE][T0][#0] interpreter = Some("/lib/ld-musl-x86_64.so.1")
[2021-03-20T01:01:23.770Z][TRACE][T0][#0] ldso_path = "/lib/ld-musl-x86_64.so.1"
[2021-03-20T01:01:23.770Z][DEBUG][T0][#0] lookup_inode: cwd: "/", path: "/lib/ld-musl-x86_64.so.1"
[2021-03-20T01:01:23.778Z][TRACE][T0][#0] interpreter = None
[2021-03-20T01:01:23.778Z][DEBUG][T0][#0] allocated rsrv addr is 0x7ffb8703d000, len is 0x9496000
[2021-03-20T01:01:23.778Z][TRACE][T0][#0] entrophy size = 19475456
[2021-03-20T01:01:23.826Z][TRACE][T0][#0] elf range = VMRange { start: 0x7ffb87200000, end: 0x7ffb87657000, size: 0x457000 }
[2021-03-20T01:01:23.879Z][TRACE][T0][#0] elf range = VMRange { start: 0x7ffb87800000, end: 0x7ffb87c96000, size: 0x496000 }
[2021-03-20T01:01:23.879Z][TRACE][T0][#0] entrophy size = 1048576
[2021-03-20T01:01:23.879Z][TRACE][T0][#0] heap range = VMRange { start: 0x7ffb87cf4000, end: 0x7ffb89cf4000, size: 0x2000000 }
[2021-03-20T01:01:23.879Z][TRACE][T0][#0] entrophy size = 1048576
[2021-03-20T01:01:23.879Z][TRACE][T0][#0] stack range = VMRange { start: 0x7ffb89d44000, end: 0x7ffb8a144000, size: 0x400000 }
[2021-03-20T01:01:23.879Z][ INFO][T0][#0] Process created: elf = /bin/init, pid = 1
[2021-03-20T01:01:23.880Z][TRACE][T1][#1][HandleException] Syscall { num = HandleException, info = 0x7ffb82578ed0, fpregs = 0x7ffb82578c70, context = 0x724fe40fe11cc119 }
[2021-03-20T01:01:23.880Z][DEBUG][T1][#1][HandleException] Handle signal: signal: FaultSignal { num: SigNum (#11 = SIGSEGV), code: 2, addr: Some(140718287196160) }, action: Dfl
[2021-03-20T01:01:23.880Z][TRACE][T1][#1][HandleException] Retval = 0
[2021-03-20T01:01:23.880Z][ INFO][T1][#1][HandleException] Thread killed: tid = 1, signum = SigNum (#11 = SIGSEGV)
[ERROR] occlum-pal: The init process exit with code: 139 (line 57, file src/pal_api.c)
[ERROR] occlum-pal: Failed to run the init process: EINVAL (line 116, file src/pal_api.c)
[ERROR] occlum-pal: Failed to do ECall: occlum_ecall_broadcast_interrupts with error code 0x2002: Invalid enclave identification. (line 27, file src/pal_interrupt_thread.c)
```

logs in occlum gdb,

```
(gdb) set environment OCCLUM_LOG_LEVEL=trace
(gdb) run  
Starting program: /root/demos/hello_c/occlum_instance/build/bin/occlum-run /bin/hello_world
warning: Error disabling address space randomization: Operation not permitted
detect urts is loaded, initializing
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
detect urts is loaded, initializing
add-symbol-file '/root/demos/hello_c/occlum_instance/build/lib/libocclum-libos.signed.so' 0x7f984001ba80 -s .interp 0x7f9840000270  -s .note.gnu.build-id 0x7f984000028c  -s .gnu.hash 0x7f98400002b0  -s .dynsym 0x7f98400002e8  -s .dynstr 0x7f9840000390  -s .gnu.version 0x7f984000040a  -s .gnu.version_d 0x7f9840000418  -s .rela.dyn 0x7f9840000450  -s .plt 0x7f984001ba50  -s .plt.got 0x7f984001ba60  -s .nipx 0x7f98401a00c0  -s .rodata 0x7f98401a0ac0  -s .niprod 0x7f98401c3940  -s .builtin_config 0x7f98401c4200  -s .eh_frame_hdr 0x7f98401c4230  -s .eh_frame 0x7f98401cc9f8  -s .gcc_except_table 0x7f98401fde74  -s .tbss 0x7f984040bbc0  -s .init_array 0x7f984040bbc0  -s .fini_array 0x7f984040bbc8  -s .data.rel.ro 0x7f984040bc20  -s .dynamic 0x7f984041bbf8  -s .got 0x7f984041bda8  -s .data 0x7f984041d000  -s .nipd 0x7f984041d350  -s .bss 0x7f984041d360 
[New Thread 0x7f9861bc7700 (LWP 431)]
[New Thread 0x7f98613c6700 (LWP 432)]
[2021-03-20T00:55:54.887Z][TRACE][T0][#0] env_checked from env untrusted: []
[2021-03-20T00:55:54.890Z][TRACE][T0][#0] env_merged = ["OCCLUM=yes"]  (default env and untrusted env)
[2021-03-20T00:55:54.892Z][DEBUG][T0][#0] lookup_inode: cwd: "/", path: "/bin/init"
[2021-03-20T00:55:55.025Z][TRACE][T0][#0] PT_INTERP offset = 568, count = 25
[2021-03-20T00:55:55.027Z][TRACE][T0][#0] interpreter = Some("/lib/ld-musl-x86_64.so.1")
[2021-03-20T00:55:55.029Z][TRACE][T0][#0] ldso_path = "/lib/ld-musl-x86_64.so.1"
[2021-03-20T00:55:55.031Z][DEBUG][T0][#0] lookup_inode: cwd: "/", path: "/lib/ld-musl-x86_64.so.1"
[2021-03-20T00:55:55.057Z][TRACE][T0][#0] interpreter = None
[2021-03-20T00:55:55.061Z][DEBUG][T0][#0] allocated rsrv addr is 0x7f984703d000, len is 0x9496000
[2021-03-20T00:55:55.063Z][TRACE][T0][#0] entrophy size = 19475456
[2021-03-20T00:55:55.263Z][TRACE][T0][#0] elf range = VMRange { start: 0x7f9847e00000, end: 0x7f9848257000, size: 0x457000 }
[2021-03-20T00:55:55.567Z][TRACE][T0][#0] elf range = VMRange { start: 0x7f9848400000, end: 0x7f9848896000, size: 0x496000 }
[2021-03-20T00:55:55.569Z][TRACE][T0][#0] entrophy size = 1048576
[2021-03-20T00:55:55.570Z][TRACE][T0][#0] heap range = VMRange { start: 0x7f98488b9000, end: 0x7f984a8b9000, size: 0x2000000 }
[2021-03-20T00:55:55.572Z][TRACE][T0][#0] entrophy size = 1048576
[2021-03-20T00:55:55.574Z][TRACE][T0][#0] stack range = VMRange { start: 0x7f984a918000, end: 0x7f984ad18000, size: 0x400000 }
add-symbol-file 'image/lib/ld-musl-x86_64.so.1' 0x7f98484142f0 -s .hash 0x7f9848400190  -s .gnu.hash 0x7f9848402c48  -s .dynsym 0x7f9848405f08  -s .dynstr 0x7f984840fe80  -s .rela.dyn 0x7f9848413bb8  -s .rela.plt 0x7f9848414218  -s .plt 0x7f9848414290  -s .rodata 0x7f984845d700  -s .eh_frame_hdr 0x7f98484912c8  -s .eh_frame 0x7f98484912f8  -s .data.rel.ro 0x7f9848691c40  -s .dynamic 0x7f9848691e20  -s .got 0x7f9848691f70  -s .got.plt 0x7f9848692000  -s .data 0x7f9848692040  -s .bss 0x7f98486924a0 
Function "_dl_map_object" not defined.
readelf: Error: 'image/bin/init': No such file
[2021-03-20T00:55:55.654Z][ INFO][T0][#0] Process created: elf = /bin/init, pid = 1

Thread 1 "occlum-run" received signal SIGSEGV, Segmentation fault.
0x00007f9848459c05 in _dlstart ()

```

**Comment by StanPlatinum** [comment]
The most weird thing is that the occlum gdb cannot find the `/bin/init`. So I suggest that we check on the output of `OCCLUM_LOG_LEVEL=trace occlum run /bin/hello_world`, aka., the first log I just posted...

**Comment by guzongmin** [comment]
Are you writing your own init application?

**Comment by StanPlatinum** [comment]
No... I used the `occlum init` command.

**Comment by guzongmin** [comment]
Can you please share your code? The CI is using exactly the same code, and it works well.

**Comment by StanPlatinum** [comment]
I was using the code at `/root/demos/hello_c` inside the container, which is

```
root@658cfea52ad3:~/demos/hello_c# cat hello_world.c 
#include <stdio.h>

int main() {
    printf("Hello World\n");
    return 0;
}
```

**Comment by liqinggd** [comment]
> I was using the code at `/root/demos/hello_c` inside the container, which is
> 
> ```
> root@658cfea52ad3:~/demos/hello_c# cat hello_world.c 
> #include <stdio.h>
> 
> int main() {
>     printf("Hello World\n");
>     return 0;
> }
> ```

I strongly suspect that the `RDFSBASE` instructions are not working in your kernel.

**Comment by StanPlatinum** [comment]
@liqinggd Thanks! Agree with you. The fsgsbase patch seems not working on my box (NUC7PJYH) machine (with Linux 5.4.0-67 installed).

However, the same error was found on my box machine with Linux 5.9.16, which has a fsgsbase patch naturally...

FYI, there was no such error on my another Xeon machine with Linux 5.9.0 kernel.

I'm starting to wonder if my box machine doesn't support this fsgsbase instruction family... but I've tested Graphene-SGX on my box machine, which worked fine (on Linux 5.9)...

Will take a deeper look at it.

**Comment by StanPlatinum** [comment]
Is there a way I can know whether the [enable_rdfsbase kernel module](https://github.com/occlum/enable_rdfsbase) is working? Any output would be very good for the driver I guess...

Thanks!

**Comment by guzongmin** [comment]
@StanPlatinum Please reference the code below
https://github.com/occlum/occlum/issues/371

**Comment by StanPlatinum** [comment]
I wrote an enablement checking program to verify if my box machine supports the fsgsbase. It turned out my box machine supports it well. So here to clarify that this may not a fsgsbase problem.

Is there any possibility that this is a bus error? Since the error message I got is `Failed to do ECall: occlum_ecall_broadcast_interrupts with error code 0x2002: Invalid enclave identification. (line 27, file src/pal_interrupt_thread.c)`
